File: 0121-best-time-to-buy-and-sell-stock/0121-best-time-to-buy-and-sell-stock.cpp
#include <iostream>
#include <algorithm> // for std::max

using std::max;
using std::min;

class Solution {
public:
    int maxProfit(vector<int>& prices) {

        int profit = 0;

        int n = prices.size();

        if(n<=1){
            return 0;
        }

        int previous = prices[0];

        for(int i=0; i<n; i++){
            
            profit = max(profit, prices[i] - previous);
            previous = min(previous, prices[i]);
            
        }

        return profit;
        
    }
};

File: 0121-best-time-to-buy-and-sell-stock/0121-best-time-to-buy-and-sell-stock.py
class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        minPrice = prices[0]
        profit = 0

        for i in range(1, len(prices)):
            price = prices[i]
            if minPrice > price:
                minPrice = price
            else:
                profit = max(profit, price - minPrice)

        return profit
        